docs/DeveloperGuide.adoc

```diff
@@ -153,62 +153,140 @@ Classes used by multiple components are in the `seedu.addressbook.commons` packa
+When an `UndoableCommand` is executed or undone, it is added to the `commandHistory` or `undoHistory` of ModelManager` respectively.
+Therefore, an `UndoableCommand` should not be undone before it was executed or redone before it was undone.
+
+To defend against improper undoing or redoing, `UndoableCommand` contains an inner class, the enumeration `UndoableCommandState` which allows an `UndoableCommand` to have its state set to `UNDOABLE`, `REDOABLE`, and `PRE_EXECUTION`.
+Before a command is undone or redone, the command's state is checked for validity.
+These states are only changed when a `Command#execute(Model model)` or `UndoableCommand#undo(Model model)` has successfully executed.
+Thus, this lowers the likelihood of bugs in undo/redo.
+
+Lastly, undo/redo is initiated when user input creates an `UndoCommand` or `RedoCommand`. When either of them are executed, they respectively get the last
+executed or undone command from the `CommandHistory` in `ModelManager`. As the retrieved command is an instance of `UndoableCommand`, an attempt will be made to execute `UndoableCommand#undo(Model model)` or `UndoableComman#redo(Model model)`.
+If it is successful, undo/redo is succesful. Otherwise, an error message is shown.
+
+This is the mechanism of undo/redo, from start to end. Below is an example of a usage scenario that illustrates how undo/redo works.
 
 [NOTE]
-If the `currentStatePointer` is at index `addressBookStateList.size() - 1`, pointing to the latest address book state, then there are no undone address book states to restore. The `redo` command uses `Model#canRedoAddressBook()` to check if this is the case. If so, it will return an error to the user rather than attempting to perform the redo.
+If a command fails to execute correctly, it will not call `Model#addExecutedCommand()` or `Model#addUndoneCommand()`, thus preventing the command from being undone or redone.
 
-Step 5. The user then decides to execute the command `list`. Commands that do not modify the address book, such as `list`, will usually not call `Model#commitAddressBook()`, `Model#undoAddressBook()` or `Model#redoAddressBook()`. Thus, the `addressBookStateList` remains unchanged.
 
-image::UndoRedoState4.png[]
+The sequence diagram below shows how an undo command works to undo a `ClearCommand`:
+
+image::UndoSequenceDiagram.png[]
 
-Step 6. The user executes `clear`, which calls `Model#commitAddressBook()`. Since the `currentStatePointer` is not pointing at the end of the `addressBookStateList`, all address book states after the `currentStatePointer` will be purged. We designed it this way because it no longer makes sense to redo the `add n/David ...` command. This is the behavior that most modern desktop applications follow.
 
-image::UndoRedoState5.png[]
+After this is done, if a `redo` command was executed, the `ClearCommand` would simply be executed again.
 
 The following activity diagram summarizes what happens when a user executes a new command:
```

Comment:
The "if-else" conditional branch could have more specific conditions; such as "user enters valid command".